### PR TITLE
Target Allocator Support for Native Prometheus Receiver

### PR DESCRIPTION
**Pull request by okankoAMZ**
# Description of the issue
We have a new tool called Target Allocator, which provides scrape configs. These changes allows telegraf based prometheus receiver to pull scrape configs from Target Allocator endpoint.

# Description of changes
* Added Target Allocator Manager(TAM), based on contrib version but with a wrapper 
* TAM is now included and checked during initial and reload configs to make sure TA is enabled or disabled.
* When TA is enabled the TAM will provide the prometheus config to telegraf prometheus receiver directly. (without a file read)

# License
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

# Tests
* Added some unit tests which pass
* Manually tested where I can see metrics on CloudWatch when Agent is deployed. 

CloudWatch Metrics ScreenShot

# Requirements
_Before commit the code, please do the following steps._
1. Run `make fmt` and `make fmt-sh`
2. Run `make lint`

### CODE REVIEW

File: plugins/inputs/prometheus/start.go [hunk 1]
```diff
@@ -258,8 +332,9 @@ func Start(configFilePath string, receiver storage.Appendable, shutDownChan chan
 				default:
 				}
 
+				<-taManager.taReadyCh
```
Comment (jefchien) [inline plugins/inputs/prometheus/start.go:335]: Will block if manager is not started. Check if the target manager is enabled.

File: plugins/inputs/prometheus/target_allocator.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+	tam.smLiveCh = make(chan struct{}, 1)
+	tam.dmLiveCh = make(chan struct{}, 1)
+	tam.taReadyCh = make(chan struct{}, 1)
+	err := tam.loadConfig(filename)
+	if err != nil {
+		level.Error(logger).Log("msg", "Error loading config", "err", err)
+		return &tam
+	}
+	tam.host = nil
+	tam.loadManager()
```
Comment (jefchien) [inline plugins/inputs/prometheus/target_allocator.go:106]: `tam.config.TargetAllocator` can be nil at this point, which would panic.

File: plugins/inputs/prometheus/start.go [hunk 3]
```diff
@@ -94,6 +136,7 @@ func Start(configFilePath string, receiver storage.Appendable, shutDownChan chan
 	cfg := struct {
 		configFile    string
 		promlogConfig promlog.Config
+		taConfig      tamanager.Config
```
Comment (jefchien) [inline plugins/inputs/prometheus/start.go:139]: Not used.

File: plugins/inputs/prometheus/target_allocator.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+	if tam.config != nil {
+		tam.enabled = (tam.config.TargetAllocator != nil) && isPodNameAvailable()
+	}
+	return &tam
+}
+func (tam *TargetAllocatorManager) loadManager() {
+	receiverSettings := receiver.Settings{
+		ID: component.MustNewID(strings.ReplaceAll(tam.config.TargetAllocator.CollectorID, "-", "_")),
+		TelemetrySettings: component.TelemetrySettings{
+			Logger:         createLogger(zapcore.DebugLevel),
```
[the remark this answers is not in the document]
Comment (okankoAMZ) [reply plugins/inputs/prometheus/target_allocator.go:115]: The otel TA Manager requires zap logger so we have to create a new one

**Review by musa-asad — COMMENTED** [review]
lgtm

**Pull request merged**